src/main/java/gregtech/common/metatileentities/storage/CraftingRecipeLogic.java

```diff
@@ -0,0 +1,198 @@
+            int remainingAmount = itemStack.getCount() - itemSources.insertItem(stackKey, itemStack.getCount(), false, IItemList.InsertMode.HIGHEST_PRIORITY);
+            if (remainingAmount > 0) {
+                itemStack.setCount(remainingAmount);
+                player.addItemStackToInventory(itemStack);
+                if (itemStack.getCount() > 0) {
+                    player.dropItem(itemStack, false, false);
+                }
+            }
+        }
+    }
+
+    public void handleItemCraft(ItemStack itemStack, EntityPlayer player, boolean simulate) {
+        itemStack.onCrafting(world, player, 1);
+        itemStack.getItem().onCreated(itemStack, world, player);
+        if (!simulate) {
+            //if we're not simulated, fire the event, unlock recipe and add crafted items, and play sounds
+            FMLCommonHandler.instance().firePlayerCraftingEvent(player, itemStack, inventoryCrafting);
+
+            if (cachedRecipe != null && !cachedRecipe.isDynamic()) {
+                player.unlockRecipes(Lists.newArrayList(cachedRecipe));
+            }
+            if (cachedRecipe != null) {
+                ItemStack resultStack = cachedRecipe.getCraftingResult(inventoryCrafting);
+                this.itemsCrafted += resultStack.getCount();
+                recipeMemory.notifyRecipePerformed(craftingGrid, resultStack);
+            }
+        }
+    }
+
+    public void refreshOutputSlot() {
+        ItemStack itemStack = ItemStack.EMPTY;
+        if (cachedRecipe != null) {
+            itemStack = cachedRecipe.getCraftingResult(inventoryCrafting);
+        }
+        this.craftingResultInventory.setInventorySlotContents(0, itemStack);
+    }
+
+    public boolean isRecipeValid() {
+        return cachedRecipeData.attemptMatchRecipe() == 511;
+    }
+
+    private void updateCurrentRecipe() {
+        if (!cachedRecipeData.matches(inventoryCrafting, world)) {
+            IRecipe newRecipe = CraftingManager.findMatchingRecipe(inventoryCrafting, world);
+            this.cachedRecipe = newRecipe;
+            ItemStack resultStack = ItemStack.EMPTY;
+            if (newRecipe != null) {
+                resultStack = newRecipe.getCraftingResult(inventoryCrafting);
+            }
+            this.craftingResultInventory.setInventorySlotContents(0, resultStack);
+            this.cachedRecipeData.setRecipe(newRecipe);
+        }
+    }
+
+    public void update() {
+        //update item sources every tick for fast tinting updates
+        itemSources.update();
+        tintLocation = getCachedRecipeData().attemptMatchRecipe();
+        if (hasCraftingGridUpdated()) {
+            updateCurrentRecipe();
+        }
+    }
+
+    public short getTintLocations() {
+        return tintLocation;
+    }
+
+    public void checkNeighbourInventories(BlockPos blockPos) {
+        for (EnumFacing side : EnumFacing.VALUES) {
+            TileItemSource itemSource = new TileItemSource(world, blockPos, side);
```

Comment:
Can this TileItemSource creation be short circuited by checking if the block on the `side` face has the itemHandler capability?